### PR TITLE
add autofix option

### PR DESCRIPTION
**Pull request by zeitlinger**
- for 1 rule in the beginning, to see if this is a good idea
- github.com/miracl/conflate has been forked - I'll create a PR upstream if this PR is accpeted - see https://github.com/miracl/conflate/compare/master...zeitlinger:conflate:merge_slice

### CODE REVIEW

**Comment by rgeyer** [comment]
This looks great. It looks like you implemented auto-fix on the `template-on-time-change-reload-rule`?

There are a few things which we should consider changing before adding this functionality.

1. The rules are intended to operate with as little context as possible. I.E. They would be given everything they need in parameters, such as a dashboard and panel reference for a panel rule. Rules should *not* require data about configuration as that's too much context.

I'd like to see the rules return a `func` reference in the `Result` struct (or something similar) which could then be executed by the caller. This would allow for the configuration to opt-out of auto-fix for a subset of rules.

2. Because of 1 above, rules would need to be able to return more than a single result. This was an early design decision, where I assumed that rules would only ever find one issue and report it, but in practice some rules can find multiple issues. As such, the interface for rules should be changed to allow for multiple `Result` from the same rule.

If we implement 1 as I suggested (a "fix" func on the `Result`) this would allow the caller/[ResultSet](https://github.com/grafana/dashboard-linter/blob/b1f5eb2cca53b30525eeca68d65e5ba017e90df2/lint/results.go#L59-L65) (which has all of the config context) to decide whether to execute auto-fix or not.

3. Many of these can not be "fixed" as they'd require the user to make informed decisions about their dashboard. For those, presumably we just wouldn't implement auto-fix.

**Comment by zeitlinger** [comment]
I think I get the idea - it's a good separation of concerns.

I've written just one method to make sure this is going into the right direction. Is that what you have in mind?

```go
func (f TargetRuleFunc) Lint(d *Dashboard, s *ResultSet) {
	for pi, p := range d.GetPanels() {
		p := p // capture loop variable
		for ti, t := range p.Targets {
			t := t // capture loop variable
			results := f.fn(*d, p, t)
			var rr []Result

			for _, r := range results.Results {
				var fix func(*Dashboard)
				if r.Fix != nil {
					fix = func(dashboard *Dashboard) {
						r.Fix(*d, p, &t)
						p.Targets[ti] = t
						d.Panels[pi] = p
					}
				}
				rr = append(rr, Result{
					Severity: r.Severity,
					Message:  r.Message,
					Fix:      fix,
				})
			}
			s.AddResult(ResultContext{
				Result:    RuleResults{rr},
				Rule:      f,
				Dashboard: d,
				Panel:     &p,
				Target:    &t,
			})
		}
	}
}
```

**Comment by rgeyer** [comment]
> I think I get the idea - it's a good separation of concerns.
> 
> I've written just one method to make sure this is going into the right direction. Is that what you have in mind?
> 
> ```go
> func (f TargetRuleFunc) Lint(d *Dashboard, s *ResultSet) {
> 	for pi, p := range d.GetPanels() {
> 		p := p // capture loop variable
> 		for ti, t := range p.Targets {
> 			t := t // capture loop variable
> 			results := f.fn(*d, p, t)
> 			var rr []Result
> 
> 			for _, r := range results.Results {
> 				var fix func(*Dashboard)
> 				if r.Fix != nil {
> 					fix = func(dashboard *Dashboard) {
> 						r.Fix(*d, p, &t)
> 						p.Targets[ti] = t
> 						d.Panels[pi] = p
> 					}
> 				}
> 				rr = append(rr, Result{
> 					Severity: r.Severity,
> 					Message:  r.Message,
> 					Fix:      fix,
> 				})
> 			}
> 			s.AddResult(ResultContext{
> 				Result:    RuleResults{rr},
> 				Rule:      f,
> 				Dashboard: d,
> 				Panel:     &p,
> 				Target:    &t,
> 			})
> 		}
> 	}
> }
> ```

Yes! This is exactly right. This would allow us to conditionally execute the fix, and allow one rule to provide multiple results/fixes.

**Comment by zeitlinger** [comment]
@rgeyer finally found the time - please have a look :smile:

**Review by rgeyer — APPROVED** [review]
Great work! Thank you for this contribution!

This PR is quite large, so I won't ask that you add it here, but.. Can you add some brief documentation on the autofix CLI flag, and what the outcome will be?

This should be in the `docs/` path. Specifically, it would be useful to add detail to `docs/rules/template-on-time-change-reload-rule.md` which explains what happens if autofix is executed for this rule.

Thank you again! 🎉 🚀

**Comment by tomwilkie** [comment]
Hi @zeitlinger & @rgeyer - I reverted this due to https://github.com/grafana/dashboard-linter/issues/136, thought it best to revert first and ask questions later.  Please do get this merged again ASAP, once we've figure out what to do about the `replace`.  Thanks!

**Comment by zeitlinger** [comment]
> Hi @zeitlinger & @rgeyer - I reverted this due to #136, thought it best to revert first and ask questions later. Please do get this merged again ASAP, once we've figure out what to do about the `replace`. Thanks!

Yes, absolutely. 
I'll check out the problem now.

File: lint/lint.go [hunk 1]
```diff
@@ -14,50 +14,48 @@ const (
...
-		Multi      bool             `json:"multi"`
-		AllValue   string           `json:"allValue"`
-		Current    TemplateValue    `json:"current"`
-		Options    []TemplateOption `json:"options"`
-		Refresh    int              `json:"refresh"`
+		Name       string             `json:"name"`
+		Label      string             `json:"label"`
+		Type       string             `json:"type"`
+		Query      interface{}        `json:"query"`
+		Datasource interface{}        `json:"datasource,omitempty"`
```
[the remark this answers is not in the document]
Comment (zeitlinger) [reply lint/lint.go:52]: it was a necessity actually - to keep the marshall/unmarshall symetric

**Comment by zeitlinger** [comment]
> This should be in the `docs/` path. Specifically, it would be useful to add detail to `docs/rules/template-on-time-change-reload-rule.md` which explains what happens if autofix is executed for this rule.

it's in https://github.com/grafana/dashboard-linter/pull/138/commits/c9518b23724f43363ecce0ef0548ca9216b2d266, which is part of https://github.com/grafana/dashboard-linter/pull/138

**Comment by zeitlinger** [comment]
> Yes, absolutely. I'll check out the problem now.

Here's the new PR: https://github.com/grafana/dashboard-linter/pull/138

**Pull request merged**